presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizationManager.java

```diff
@@ -0,0 +1,268 @@
+
+    private final IDBI dbi;
+    private final MetadataDao metadataDao;
+    private final ShardOrganizerDao shardOrganizerDao;
+    private final ShardManager shardManager;
+
+    private final boolean enabled;
+    private final Duration discoveryInterval;
+
+    private final String currentNodeIdentifier;
+    private final ShardOrganizer organizer;
+
+    @Inject
+    public ShardOrganizationManager(@ForMetadata IDBI dbi, NodeManager nodeManager, ShardManager shardManager, ShardOrganizer organizer, StorageManagerConfig config)
+    {
+        this(dbi,
+                nodeManager.getCurrentNode().getNodeIdentifier(),
+                shardManager,
+                organizer,
+                config.isOrganizationEnabled(),
+                config.getOrganizationInterval());
+    }
+
+    public ShardOrganizationManager(
+            IDBI dbi,
+            String currentNodeIdentifier,
+            ShardManager shardManager,
+            ShardOrganizer organizer,
+            boolean enabled,
+            Duration discoveryInterval)
+    {
+        this.dbi = requireNonNull(dbi, "dbi is null");
+        this.metadataDao = onDemandDao(dbi, MetadataDao.class);
+        this.shardOrganizerDao = onDemandDao(dbi, ShardOrganizerDao.class);
+
+        this.organizer = requireNonNull(organizer, "organizer is null");
+        this.shardManager = requireNonNull(shardManager, "shardManager is null");
+        this.currentNodeIdentifier = requireNonNull(currentNodeIdentifier, "currentNodeIdentifier is null");
+
+        this.enabled = enabled;
+        this.discoveryInterval = requireNonNull(discoveryInterval, "discoveryInterval is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (!enabled) {
+            return;
+        }
+
+        if (!started.getAndSet(true)) {
+            startDiscovery();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        shutdown.set(true);
+        discoveryService.shutdown();
+    }
+
+    private void startDiscovery()
+    {
+        long startDelay = getStartDelay();
+
+        discoveryService.scheduleWithFixedDelay(() -> {
+            if (shutdown.get()) {
+                return;
+            }
+
+            try {
+                // jitter to avoid overloading database
+                long intervalSeconds = min(3600, (long) discoveryInterval.convertTo(SECONDS).getValue());
+                SECONDS.sleep(ThreadLocalRandom.current().nextLong(1, intervalSeconds));
+
+                log.info("Running shard organizer...");
+                discoverShardsAndSubmitJobs();
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error running shard organizer");
+            }
+            finally {
+                shardOrganizerDao.updateShardOrganizerLastRunTime(currentNodeIdentifier);
```

Comment:
This shouldn't update if there was an error. Otherwise, for example, a transient database will cause it to miss a run